api/src/wfl/stage.clj

```diff
@@ -0,0 +1,30 @@
+(ns wfl.stage
+  "An interface for operations on a queue-based pipeline processing stage,
+  e.g. source, executor, or sink."
+  (:require [wfl.util :as util])
+  (:import [wfl.util UserException]))
+
+(defmulti validate-or-throw
+  "Validate the `request` request."
+  (fn [request] (:name request)))
+
+(defmethod validate-or-throw :default
+  [{:keys [name] :as request}]
+  (throw (UserException. "Invalid request - unknown name"
+                         (util/make-map name request))))
+
+(defmulti queue-peek
+  "Peek the first object from the `queue`, if one exists."
+  (fn [queue] (:type queue)))
+
+(defmulti queue-pop!
```

Comment:
In an earlier [Slack discussion](https://broadinstitute.slack.com/archives/CH5CJ5FSN/p1623271960005000) I ran into a naming collision against `clojure.core` when defining a `queue` namespace with a `pop!` method.

As I thought about it, in this case a queue is a possible attribute of a pipeline processing stage.  I ended up grouping the interfaces together, avoiding the naming collision as a side effect.